### PR TITLE
Autoregressive moving average [WIP]

### PR DESCRIPTION
**Pull request by jakee417**
Implements a Hamilton ARMA(p, p-1) as referenced in [this issue](https://github.com/tensorflow/probability/issues/677).

Two tests are completed (batch shape and AR(p) equivalence), manual ARMA log_prob calculation is still WIP.

Since we need ARMA(p, p-1) to get correct shapes for `tfd.LinearGaussianStateSpaceModel`, I think the best design is to automatically pad either the `ar_coefficients` or `ma_coefficients` to meet this requirement (i.e. `ar=[1, 1, 1]`, `ma=[1, 1, 1]` becomes `ar=[1, 1, 1, 0]`, `ma=[1, 1, 1]`. Thus enabling users access to the general ARMA(p, q) formulation without too much complaining from the api. This might lead to unexpected behavior for `initial_state_prior` since this argument will need to be of shape `[batch, max(p, q + 1)]` which might be mentioned as a gotcha in the documentation.

Code was compared against [this statsmodel example](https://www.statsmodels.org/dev/examples/notebooks/generated/statespace_sarimax_internet.html) up to ARMA(2, 2) models:

``` python
import pytest
import pandas as pd
import statsmodels.api as sm
from autotsml.layers.probabilistic import ARMAStateSpaceModel
import tensorflow_probability as tfp
import tensorflow as tf
import numpy as np
import warnings

warnings.simplefilter('ignore')

tfd = tfp.distributions

import requests
from io import BytesIO
from zipfile import ZipFile

MAX_STEP = 5
E = 0.05

# Download the dataset
dk = requests.get('http://www.ssfpack.com/files/DK-data.zip').content
f = BytesIO(dk)
zipped = ZipFile(f)
df = pd.read_table(
    BytesIO(zipped.read('internet.dat')),
    skiprows=1, header=None, sep='\s+', engine='python',
    names=['internet', 'dinternet']
)
# Get the basic series
dta_full = df.dinternet[1:].values
x = dta_full.reshape(-1, 1)
dta_miss = dta_full.copy()

# Remove datapoints
missing = np.r_[6, 16, 26, 36, 46, 56, 66, 72, 73, 74, 75, 76, 86, 96] - 1
dta_miss[missing] = np.nan

optimizer = tf.keras.optimizers.Adam(learning_rate=0.1)

@pytest.mark.parametrize("p", [1, 2])
@pytest.mark.parametrize("q", [1, 2])
def test_arma_statsmodel(p, q):
    """https://www.statsmodels.org/dev/examples/notebooks/generated/statespace_sarimax_internet.html"""
    model = sm.tsa.statespace.SARIMAX(
        x, order=(p, 0, q),
        hamilton_representation=True,
        enforce_invertibility=False).fit()

    # warm start tfp since this problem is non-linear and highly sensitive to
    # the initial parameter configurations
    ar_init = model.arparams + tf.random.normal(model.arparams.shape, stddev=E)
    ma_init = model.maparams + tf.random.normal(model.maparams.shape, stddev=E)

    # instantiate the parameters for the ARMA model
    ar_params = tf.Variable(ar_init, trainable=True, name='ar')
    ma_params = tf.Variable(ma_init, trainable=True, name='ma')
    scale_params = tf.Variable(1., trainable=True, name='scale')

    def make_arma_model():
        return ARMAStateSpaceModel(
            num_timesteps=len(x),
            ar_coefficients=ar_params,
            ma_coefficients=ma_params,
            level_scale=scale_params,
            initial_state_prior=tfd.MultivariateNormalDiag(
                scale_diag=[1.] * max(p, q + 1))
        )

    def nll():
        return -make_arma_model().log_prob(x)

    @tf.function(autograph=False)
    def train_op():
        with tf.GradientTape() as tape:
            loss = nll()
        grads = tape.gradient(
            loss, make_arma_model().trainable_variables)
        optimizer.apply_gradients(
            zip(grads, make_arma_model().trainable_variables))
        return loss, make_arma_model().trainable_variables

    for step in range(MAX_STEP):
        loss, vars = train_op()
        vars = [i.numpy() for i in vars]
        if step % 20 == 0:
            print("step {}: log prob {}: vars {}".format(step, loss, vars))

    np.testing.assert_allclose(model.arparams, ar_params.numpy(), atol=0.5)
    np.testing.assert_allclose(model.maparams, ma_params.numpy(), atol=0.5)
```

But was omitted since this test does not qualify as lightweight (which was requested).

The author and devs of statsmodel both note that parameter estimation becomes unstable with higher order models, I see this begins at order 3 (i.e. ARMA(3, 3))

### CODE REVIEW

File: tensorflow_probability/python/sts/components/autoregressive_moving_average.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+                initial_state_prior=initial_state_prior,
+                name=name,
+                **linear_gaussian_ssm_kwargs)
+            self._parameters = parameters
+
+    @property
+    def order(self):
+        return self._order
+
+    @order.setter
```
Comment (davmre) [inline tensorflow_probability/python/sts/components/autoregressive_moving_average.py:198]: Is this setter needed/used anywhere? We generally don't allow mutation of public attributes on TFP objects.

**Review by davmre — COMMENTED** [review]
Hi Jake, this is great, thanks for putting it together! Your design decisions around padding seem right to me. I left a few minor comments, but overall I think the code looks to be in quite good shape.

On the annoying legal side, before we can pull this in we'll also need a CLA from you and/or your employer. Can you take a look at https://cla.developers.google.com/clas when you get a chance?

- Dave

File: tensorflow_probability/python/sts/components/autoregressive_moving_average.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+    The Hamilton autoregressive moving average model implements a
+    `tfp.distributions.LinearGaussianStateSpaceModel` with `latent_size = order`
+    and `observation_size = 1`. The latent state vector encodes the recent history
+    of the process, with the current value in the topmost dimension. At each
+    timestep, the transition sums the previous values to produce the new expected
+    value, shifts all other values down by a dimension, and adds noise to the
+    current value. This is formally encoded by the transition model:
+
+    ```
+    transition_matrix = [ phi[0], phi[1],     ..., phi[p]
```
[the remark this answers is not in the document]
Comment (jakee417) [reply tensorflow_probability/python/sts/components/autoregressive_moving_average.py:69]: `ar_coefs` fits and matches [current documentation ](https://www.tensorflow.org/probability/api_docs/python/tfp/sts/AutoregressiveStateSpaceModel)(more or less).

File: tensorflow_probability/python/sts/components/autoregressive_moving_average.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+                          0.,       0.,  ...,  1.,  0.   ]
+
+    transition_noise ~ N(loc=0., scale=diag([level_scale, 0., 0., ..., 0.]))
+    ```
+
+    The observation model simply extracts the current (topmost) value,
+    sums the previous noise and optionally adds independent noise at each step:
+
+    ```
+    observation_matrix = [1, theta[0], theta[1], ..., theta[p-1]]
```
[the remark this answers is not in the document]
Comment (jakee417) [reply tensorflow_probability/python/sts/components/autoregressive_moving_average.py:82]: Went with `ma_coefs` for consistency with [above](https://github.com/tensorflow/probability/pull/1474#discussion_r765061744).

File: tensorflow_probability/python/sts/components/autoregressive_moving_average.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+    observation_noise ~ N(loc=0, scale=observation_noise_scale)
+    ```
+
+    Models with `observation_noise_scale = 0.` are ARMA(p, p-1) processes
+    in the formal sense. Setting `observation_noise_scale` to a nonzero value
+    corresponds to a latent ARMA(p, p-1) process observed under an iid noise model.
+
+    See the [Wikipedia article](http://web.pdx.edu/~crkl/readings/Hamilton94.pdf)
```
[the remark this answers is not in the document]
Comment (jakee417) [reply tensorflow_probability/python/sts/components/autoregressive_moving_average.py:90]: Whoops! I was trying to match [this style](https://www.tensorflow.org/probability/api_docs/python/tfp/sts/Autoregressive) in `autoregressive.py`, but yes, the reference is much cleaner.

File: tensorflow_probability/python/sts/components/autoregressive_moving_average.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+                 **linear_gaussian_ssm_kwargs):
+        """ Build a state space model implementing an ARMA(p, p - 1) process.
+
+        Args:
+            num_timesteps: Scalar `int` `Tensor` number of timesteps to model
+                with this distribution.
+            ar_coefficients: `float` `Tensor` of shape `concat(batch_shape, [order])`
+                defining  the autoregressive coefficients. The coefficients are defined
+                backwards in time: `coefficients[0] * level[t] + coefficients[1] *
+                level[t-1] + ... + coefficients[order-1] * level[t-order+1]`.
```
[the remark this answers is not in the document]
Comment (jakee417) [reply tensorflow_probability/python/sts/components/autoregressive_moving_average.py]: I actually had a mistake on `ma_coefficients` as the current noise `noise[t]` does not get a `ma_coefficients`. So it should read:
```
ma_coefficients: `float` `Tensor` of shape `concat(batch_shape, [order])`
    defining  the moving average coefficients. The ma_coefficients are defined
    backwards in time: `noise[t] + ma_coefficients[0] * noise[t-1]
    + ... + ma_coefficients[order-2] * noise[t-order+1]`.
```
But otherwise cleaned the names up in the docs.

**Comment by jakee417** [comment]
@davmre I got a personal CLA on file and took care of the quick and easy documentation issues this morning before work. I will get to the `batch_shape` issue and also the `log_prob` calculation tonight!

File: tensorflow_probability/python/sts/components/autoregressive_moving_average.py [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+                raise ValueError('coefficients must have static shape.')
+
+            # if p != q + 1, then pad either of the arguments to ensure
+            # we end up fitting an ARMA(p, p-1) process.
+            if p > q + 1:
+                pad = tf.zeros([p - q - 1], dtype=dtype)
+                ma_coefficients = tf.concat([ma_coefficients, pad], axis=-1)
+            elif q + 1 > p:
+                pad = tf.zeros([q + 1 - p], dtype=dtype)
+                ar_coefficients = tf.concat([ar_coefficients, pad], axis=-1)
```
[the remark this answers is not in the document]
Comment (jakee417) [reply tensorflow_probability/python/sts/components/autoregressive_moving_average.py:171]: I was getting `dtype` issues with your approach. Conceptually, this code is a little easier for me:
```
def _pad_tensor_with_trailing_zeros(x, num_zeros):
    pad = tf.zeros_like(x, dtype=x.dtype)[..., 0:1]
    pad = tf.repeat(pad, num_zeros, axis=-1)
    return tf.concat([x, pad], axis=-1)
```
So maybe I will propose this as it passes tests. 

Also, as this padding may frustrate users because the event shape of the prior needs to match `max(p, q + 1)`, I think adding:
```
            # ensure the prior's shape matches the padded order
            prior_event_shape = tf.compat.dimension_value(
                initial_state_prior.event_shape[-1])
            if prior_event_shape != self.order:
                raise ValueError('prior event shape needs to match max(p, q + 1). '
                                 f'prior event shape: {prior_event_shape}, '
                                 f'max(p, q + 1): {self.order}.')
```
would be best

**Comment by jakee417** [comment]
Okay, should have everything resolved except `log_prob` exact calculation. I tried computing the conditional log probability from page 132 of [this](http://www.ru.ac.bd/stat/wp-content/uploads/sites/25/2019/03/504_02_Hamilton_Time-Series-Analysis.pdf) and it is not matching that of the `LinearGaussianStateSpaceModel`. I believe `ssm.log_prob(observed_time_series[..., tf.newaxis])` in the `LinearGaussianStateSpaceModel` might be computing the exact log likelihood if I am not mistaken?

File: tensorflow_probability/python/sts/components/autoregressive_moving_average.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (davmre) [reply tensorflow_probability/python/sts/components/autoregressive_moving_average.py]: Ah, right---my suggestion should have included `dtype` in the `tf.zeros([tf.rank(x) - 1, 2], dtype=x.dtype)` term. Your approach also looks correct, so doing it that way is fine by me.

Minor nit: `tf.zeros_like(x, dtype=x.dtype)[..., 0:1]` would be better as `tf.zeros_like(x[..., 0:1])`, since passing `x.dtype` to `zeros_like` is redundant, and doing the slicing inside avoids allocating an unnecessary temporary array of zeros. (though I'm sure the actual performance difference is negligible).

**Comment by davmre** [comment]
> Okay, should have everything resolved except `log_prob` exact calculation. I tried computing the conditional log probability from page 132 of [this](http://www.ru.ac.bd/stat/wp-content/uploads/sites/25/2019/03/504_02_Hamilton_Time-Series-Analysis.pdf) and it is not matching that of the `LinearGaussianStateSpaceModel`. I believe `ssm.log_prob(observed_time_series[..., tf.newaxis])` in the `LinearGaussianStateSpaceModel` might be computing the exact log likelihood if I am not mistaken?

Yes---it looks like the Hamilton formula conditions on specific values of the `epsilon` noise terms, whereas the SSM will integrate over them. I'm not an expert, but I'd imagine that knowledge of the coefficients and a long enough history of observed `y` values should be enough to identify the noise terms, so it seems like you'd expect the discrepancy to arise mostly in the first few steps, before there's history to fall back on? This would predict that the two likelihoods will tend to converge as you make the series longer, in which case you could just do that and add a small tolerance to account for the uncertainty about initial conditions.

(alternately, you could try to work out an SSM initial state prior that would imply specific previous values of `epsilon`, to get an exact correspondence, but that seems likely to be more trouble than it's worth).

**Review by davmre — APPROVED** [review]
Thanks Jake! This looks great, pending figuring out the log prob test (or just removing it if that's too much trouble).

**Comment by davmre** [comment]
I'll be out tomorrow, but this looks basically ready to pull in once you've had a chance to take a final pass, so hopefully we'll be able to do that early next week.

It looks like the bot hasn't recognized your CLA yet---one way this can happen is if you have multiple email addresses associated with your Github account, since the checks run at the level of individual commits rather than your account as a whole. Sorry for the extra hurdle, but would you mind trying to rebase this PR into a commit(s) from the address that you signed the CLA with?

**Comment by davmre** [comment]
Superseded by https://github.com/tensorflow/probability/pull/1476.
